evm/p2p/peer.py

```diff
@@ -280,48 +338,117 @@ def match_protocols(self, remote_capabilities):
+        remote peer. It is available only after the initial P2P handshake.
+        """
+        assert len(self.enabled_sub_protocols) > 0
+        if self._les_proto is None:
+            for proto in self.enabled_sub_protocols:
+                if proto.name == LESProtocol.name:
+                    self._les_proto = proto
+        return self._les_proto
+
+
+class HeadInfo:
+    def __init__(self, block_number, block_hash, total_difficulty, genesis_hash):
+        self.block_number = block_number
+        self.block_hash = block_hash
+        self.total_difficulty = total_difficulty
+        self.genesis_hash = genesis_hash
+
+
 if __name__ == "__main__":
     """
     Run geth like this to be able to do a handshake and get a Peer connected to it.
     ./build/bin/geth -vmodule p2p=4,p2p/discv5=0,eth/*=0 \
       -nodekeyhex 45a915e4d060149eb4365960e6a7a45f334393093061116b197e3240065ff2d8 \
       -port 30301 -nat none -testnet -lightserv 90
     """
+    import argparse
+    from evm.chains.ropsten import (
+        RopstenChain,
+        ROPSTEN_GENESIS_HEADER,
+    )
+    from evm.db.backends.memory import MemoryDB
+    from evm.db.backends.level import LevelDB
+    from evm.db.chain import BaseChainDB
     from evm.p2p import kademlia
-    logging.basicConfig(level=logging.DEBUG, format='%(levelname)s: %(message)s')
+    logging.basicConfig(level=logging.INFO, format='%(levelname)s: %(message)s')
     remote_pubkey = keys.PrivateKey(decode_hex(
         "0x45a915e4d060149eb4365960e6a7a45f334393093061116b197e3240065ff2d8")).public_key
     remote = kademlia.Node(remote_pubkey, kademlia.Address('127.0.0.1', 30301, 30301))
 
-    loop = asyncio.get_event_loop()
-    loop.set_debug(True)
-
-    peers = []
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-db', type=str)
+    args = parser.parse_args()
 
-    @asyncio.coroutine
-    def do_handshake():
-        try:
-            peer = yield from asyncio.wait_for(
-                handshake(remote, ecies.generate_privkey()),
-                HANDSHAKE_TIMEOUT)
-            peers.append(peer)
-        except UnreachablePeer as e:
-            print("Failed to connect to {}: {}".format(remote, e))
-        except UselessPeer:
-            print("No matching capabilities with {}".format(remote))
-        except asyncio.TimeoutError as e:
-            print("Timeout performing handshake with {}: {}".format(remote, e))
-        except Exception as e:
-            print("Unexpected error during auth/p2p handhsake with {}: {}".format(remote, e))
+    genesis_header = ROPSTEN_GENESIS_HEADER
+    if args.db is not None:
+        chaindb = BaseChainDB(LevelDB(args.db))
+    else:
+        chaindb = BaseChainDB(MemoryDB())
+    try:
+        chaindb.get_canonical_head()
+        # If the above doesn't raise CanonicalHeadNotFound it means we're reusing an existing db.
+        chain = RopstenChain(chaindb)
```

Comment:
Can you move this to an `else` clause so that it's outside of the `try` block to remove the possibility of catching an exception we don't mean to.

Reply:
:+1: